### PR TITLE
DBZ-7759 Ensure error handling of oversized events swallowed by BufferedChangeStreamCursor thread

### PR DESCRIPTION
**Pull request by ani-sha**
closes https://issues.redhat.com/browse/DBZ-7759

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java [hunk 1]
```diff
@@ -225,15 +227,23 @@ private void fetchEvents(MongoChangeStreamCursor<ChangeStreamDocument<TResult>>
...
         private Optional<ResumableChangeStreamEvent<TResult>> fetchEvent(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) {
             var beforeEventPollTime = clock.currentTimeAsInstant();
-            var document = cursor.tryNext();
+            ChangeStreamDocument<TResult> document;
+            try {
+                document = cursor.tryNext();
+            }
+            catch (MongoException e) {
+                running.set(false);
+                throw new MongoChangeStreamException("Error while fetching change stream event" + e.getMessage());
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java:236]: Do we need to use `MongoChangeStreamException`? If not, let's use another one so the original exception is added to the stack.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java [hunk 2]
```diff
@@ -297,8 +307,18 @@ public BufferingChangeStreamCursor<TResult> start() {
...
+                lastResumeToken = event.resumeToken;
+            }
+            else {
+                if (!fetcher.isRunning() && fetcher.isEmpty()) {
+                    throw new MongoChangeStreamException("Failed to fetch event from change stream cursor, cursor is closed");
+                }
+            }
+        }
+        catch (MongoException e) {
+            throw new MongoException(e.getMessage());
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java:321]: Please add original exception to the stack

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java [hunk 3]
```diff
@@ -826,6 +826,66 @@ public void shouldConsumeAllEventsFromDatabase() throws InterruptedException, IO
...
+            var updateDocument = new Document("$set", new Document("value", afterValue));
+            updateDocument(dbName, collName, new Document("_id", docId), updateDocument);
+        }
+
+        // Consume records
+        SourceRecords records = consumeRecordsByTopic(1);
+        assertThat(records.allRecordsInOrder().size()).isEqualTo(0);
+
+        // Stop the connector and check for error message
+        stopConnector(value -> assertThat(logInterceptor.containsErrorMessage("Error while reading change stream")).isTrue());
```
Comment (jpechane) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:885]: I think the error message is different

**Review by jpechane — CHANGES_REQUESTED** [review]
@ani-sha Thanks for the PR, it is a good start but needs some polish.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java (hunk 3 above)
Comment (ani-sha) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:885]: It's the error from `MongoDbStreamingChangeEventSource#readChangeStream` catch clause.

Comment (jpechane) [reply debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/MongoDbConnectorIT.java:885]: I don't see such error message there.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbStreamingChangeEventSource.java [hunk 4]
```diff
@@ -118,6 +119,10 @@ private void readChangeStream(MongoClient client, ChangeEventSourceContext conte
                 }
             }
         }
+        catch (MongoException e) {
```
Comment (ani-sha) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbStreamingChangeEventSource.java:122]: @jpechane It's here ^^

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java (hunk 2 above)
Comment (jcechace) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java]: Definitively Could use DebeziumException as a wrapper

**Comment by jcechace** [comment]
I don't like the approach where the exception is propagated via a cascade of throws up to the main streaming loop. Utilising a reference like in https://github.com/nj4x/debezium/commit/4ca19aa90d4552833c595fcf4a147b14f9b72d58 seems as a cleaner solution (except that commits forgets to mark the thread as stopped). 

Could we do something like that instead?

**Review by jcechace — CHANGES_REQUESTED** [review]
I don't like the approach where the exception is propagated via a cascade of throws up to the main streaming loop. Utilising a reference like in https://github.com/nj4x/debezium/commit/4ca19aa90d4552833c595fcf4a147b14f9b72d58 seems as a cleaner solution (except that commits forgets to mark the thread as stopped). 

Could we do something like that instead?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java [hunk 5]
```diff
@@ -198,7 +224,8 @@ public void run() {
                 noMessageIterations = 0;
                 fetchEvents(cursor);
             }
-            catch (InterruptedException e) {
+            catch (Throwable e) {
+                error.set(e);
                 throw new DebeziumException("Fetcher thread interrupted", e);
             }
```
Comment (jcechace) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java:230]: This needs to stay as two separate branches. With `InterruptedException` we want to end right away. Whereas on other exceptions we want to keep the opportunity to drain the current buffer queue. 

So we should set the error and mark the fetcher as not running.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java [hunk 6]
```diff
@@ -163,14 +165,38 @@ public boolean isRunning() {
...
         public void close() {
             running.set(false);
         }
 
         public ResumableChangeStreamEvent<TResult> poll() {
             var event = queue.poll();
-            if (event != null) {
+            if (event == null) {
+                if (hasError()) {
+                    close();
```
Comment (jcechace) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java:195]: I think the fetter should stop itself on error.

**Review by jcechace — CHANGES_REQUESTED** [review]
@ani-sha Almost there... left few small comments. 

BTW: when you are at it. The referenced commit from our user added some useful logs. Could you incorporate those as well please?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java [hunk 7]
```diff
@@ -199,8 +224,14 @@ public void run() {
                 fetchEvents(cursor);
             }
             catch (InterruptedException e) {
+                LOGGER.error("Fetcher thread interrupted", e);
                 throw new DebeziumException("Fetcher thread interrupted", e);
             }
+            catch (Throwable e) {
+                error.set(e);
+                LOGGER.error("Fetcher thread has failed", e);
+                running.set(false);
```
Comment (jcechace) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java:233]: No need, it's closed in the finally block.

**Review by jcechace — APPROVED** [review]
LGTM... tests improvements can be submitted in a new PR

**Comment by jpechane** [comment]
@ani-sha Applied, thanks!

**Pull request merged**